use embeded fftw3 as a subpackage

rainwoodman opened the pull request:
This hugely simplifies the building process of pfft. Since the new
planners are unlikely to be included in next release of FFTW in any short term, it
makes sense to just ship in a release the patched version of FFTW
that is known to work with PFFT.

This currently requires --enable-mpi on the configure scripts. threading support is untested.

rainwoodman committed:
use embeded fftw3 as a subpackage

This hugely simplifies the building process of pfft. Since the new
planners are unlikely to be included in next release of FFTW, it
makes sense to just ship in a release the patched version of FFTW
that is known to work with PFFT.

touch fftw3/Changelog

mpip commented:
Thank you for the patch. I will have a look on it. However, it might be a bit more difficult, since PFFT is also a submodule of PNFFT and ScaFaCoS. ScaFaCoS comes with its own FFTW clone -- but there we must prefix all FFTW functions with fcs_fftw_ in order to avoid linking errors.

I have to think of a nice approach that allows PFFT to come with its own FFTW and PNFFT to come with its own FFTW and PFFT. The solution must be compatible with ScaFaCoS, that comes with its own FFTW, PFFT and PNFFT. In addition, it should be possible that the user provides its own versions of the dependencies via CFLAGS and LDFLAGS.

mpip commented:
Any ideas are appreciated.

rainwoodman commented:
May I ask why PFFT has a FFTW, PNFFT has another FFTW, and ScaFaCoS has a third FFTW? What is the difference between these FFTWs?

ghisvail commented:
FYI, a requirement on a patched version of FFTW would seriously hinder any attempts to package these pieces of software for Linux.

rainwoodman commented:
on the PFFT side, an alternative is to give FFTW the ability to accept additional planners. That may be easier to get accepted than, supplying with new planners that FFTW upstream doesn't make use of.

What customization to FFTW does PNFFT and ScaFaCoS require?

rainwoodman commented:
I agree with @ghisvail  -- was bitten by this before.

rainwoodman committed:
Better omp support?

rainwoodman committed:
Better omp support

rainwoodman committed:
avoid backslash

rainwoodman committed:
update fftw3.

rainwoodman committed:
bump version

rainwoodman committed:
bump version and embedded fftw

rainwoodman committed:
use fftw 3.3.7

rainwoodman committed:
add 2d on 2d test case.

rainwoodman committed:
update 2don2d

rainwoodman committed:
add generic remap_nd interface.

rainwoodman committed:
Add some comments in the code.

rainwoodman committed:
add a 2dto1d remapping that compiles.

rainwoodman committed:
add interface of 2dto1d in remap.c

rainwoodman committed:
At least now the test case runs.

But the result is garbage.

rainwoodman committed:
fix the 2don2d test case on 1x1.

rainwoodman committed:
update tests.

rainwoodman committed:
c2c is correct but c of r2c is wrong.

however the round trip is correct.
So looks like we are doing some transpose incorrectly but consistently.

rainwoodman committed:
Actually c2c also doesn't work on 4x4 with 1x2.

rainwoodman committed:
add gather and print helpers.

rainwoodman committed:
rename members of fftwplan for better understanding.

rainwoodman committed:
commit c2c 2don2d test.

rainwoodman committed:
remove noisy message from api.

rainwoodman committed:
add the third transpose that breaks the roundtrip.

But without it the single tranform was wrong anyways.

rainwoodman committed:
c2c_3don3d prints the array for easier consistency checks.

Commit:
Minor: fix calculation of plain index (only used for generation of init data, not part of the pfft kernel)

Commit:
Minor: cleanup 2dto1d remap

Commit:
Minor: edit comment

Commit:
bugfix: fix order of in and out arrays

rainwoodman committed:
protect critical remap_routines with needs_remap.

The old head breaks pfft-python everywhere with a crash in
local_size_remap_nd_transposed

rainwoodman committed:
remove a few printfs

rainwoodman committed:
follow convention to write to input if DESTROY_INPUT is set.

This fixed all cases but the PADDED r2C and likely the shifted transforms.

The problem with the padded r2c transforms is that somehow the last axis local_ni
is not padded. It should be a trivial fix?

I am lost here, because the size is actually computed outside the remap routines
by partrafo.c The routines looked generic enough to support any decompositions,
including 2d on 2d and 3d on 3d.

rainwoodman committed:
clean up the code.

rainwoodman committed:
roughly follow the heuristics in remap_3dto2d.c

It appears that we shall prefer out of place global_transp
as much as possible.

The FFTW_PRESERVE_INPUT flag is added to signify that the transform
must be preserving input; it should have been added automatically.

Also expanding the code branches to make it easier to parse
what is actually done in different cases.

rainwoodman committed:
update the name to label the 2don2d release.
